lib/MultiTopicsConsumerImpl.cc

```diff
@@ -899,50 +884,50 @@ std::shared_ptr<TopicName> MultiTopicsConsumerImpl::topicNamesValid(const std::v
     return topicNamePtr;
 }
 
-void MultiTopicsConsumerImpl::seekAsync(const MessageId& msgId, ResultCallback callback) {
-    callback(ResultOperationNotSupported);
-}
-
-void MultiTopicsConsumerImpl::seekAsync(uint64_t timestamp, ResultCallback callback) {
-    if (state_ != Ready) {
-        callback(ResultAlreadyClosed);
-        return;
-    }
-
+void MultiTopicsConsumerImpl::beforeSeek() {
     duringSeek_.store(true, std::memory_order_release);
     consumers_.forEachValue([](const ConsumerImplPtr& consumer) { consumer->pauseMessageListener(); });
     unAckedMessageTrackerPtr_->clear();
     incomingMessages_.clear();
     incomingMessagesSize_ = 0L;
+}
+
+void MultiTopicsConsumerImpl::afterSeek() {
+    duringSeek_.store(false, std::memory_order_release);
+    auto self = get_shared_this_ptr();
+    listenerExecutor_->postWork([this, self] {
+        consumers_.forEachValue([](const ConsumerImplPtr& consumer) { consumer->resumeMessageListener(); });
+    });
+}
+
+void MultiTopicsConsumerImpl::seekAsync(const MessageId& msgId, ResultCallback callback) {
+    if (msgId == MessageId::earliest() || msgId == MessageId::latest()) {
+        return seekAllAsync(msgId, callback);
+    }
+
+    auto optConsumer = consumers_.find(msgId.getTopicName());
+    if (!optConsumer) {
+        callback(ResultOperationNotSupported);
```

Comment:
I think they are the same. An empty topic is not supported because there is no internal consumer that subscribes an empty topic.

> we add a method to set the topicName in MessageIdBuilder?

Yes. What you suggested is the `TopicMessageId#create` method here: https://github.com/apache/pulsar/blob/3534cdd06f56cb6d7c072b8d70f5b729bc692343/pulsar-client-api/src/main/java/org/apache/pulsar/client/api/TopicMessageId.java#L44

It's another feature that allows users to specify the topic name to seek or acknowledge for a MessageId not returned by `Message.getMessageId`.

Reply:
For now, I will add an error log here because C++ client cannot expose error message via the API currently.